File: untracked/solver.py

import json
from itertools import combinations

def solve():
    with open('data.json', 'r', encoding='utf-8') as f:
        data = json.load(f)
    with open('inventory.json', 'r', encoding='utf-8') as f:
        inventory = json.load(f)

    weapons = data['weapons']
    locations = data['alluvium']
    attributes = ['Strength', 'Agility', 'Will', 'Intellect', 'Main Attribute']

    # Assign priorities based on projectgoal.txt
    for w in weapons:
        inv = inventory.get(w['name'], {'weapon_count': 0, 'essence_count': 0, 'desired': False})
        wc = inv.get('weapon_count', 0)
        ec = inv.get('essence_count', 0)
        desired = inv.get('desired', False)
        
        if desired:
            priority = 5
        elif ec == 0 and wc > 0:
            priority = 4
        elif ec == 0 and wc == 0:
            priority = 3
        elif ec > 0 and wc > ec:
            priority = 2
        elif ec > 0 and wc == ec:
            priority = 1
        else:
            priority = 1
        
        w['priority'] = priority

    results = []

    attr_combos = list(combinations(attributes, 3))

    for loc_name, loc_data in locations.items():
        loc_sec = set(loc_data['secondary'])
        loc_skill = set(loc_data['skill'])
        
        # 16 possible selections: 8 Secondary Stats, 8 Skill Stats
        selections = []
        for s in loc_sec:
            selections.append(('secondary', s))
        for k in loc_skill:
            selections.append(('skill', k))

        for sel_type, sel_val in selections:
            for attr_sel in attr_combos:
                valid_outcomes = []
                
                for w in weapons:
                    if w['priority'] <= 1:
                        continue
                        
                    if w['attr_stat'] not in attr_sel:
                        continue
                    
                    if sel_type == 'secondary':
                        # If we pick a Secondary, the weapon must match that secondary 
                        # AND its skill must be in the location's skill pool.
                        if w['sec_stat'] == sel_val and w['skill_stat'] in loc_skill:
                            valid_outcomes.append(w)
                    else:
                        # If we pick a Skill, the weapon must match that skill
                        # AND its secondary must be in the location's secondary pool.
                        if w['skill_stat'] == sel_val and w['sec_stat'] in loc_sec:
                            valid_outcomes.append(w)
                
                if valid_outcomes:
                    score = sum(w['priority'] for w in valid_outcomes)
                    distinct_attrs = len(set(w['attr_stat'] for w in valid_outcomes))
                    results.append({
                        'location': loc_name,
                        'selection': f"{'/'.join(attr_sel)} + {sel_val}",
                        'valid_count': len(valid_outcomes),
                        'score': score,
                        'distinct_attrs': distinct_attrs,
                        'outcomes': [w['name'] for w in valid_outcomes],
                        'priorities': [w['priority'] for w in valid_outcomes]
                    })

    # Ranking:
    # 1. valid_count (touches most unobtained)
    # 2. score (sum of priorities)
    # 3. distinct_attrs (tiebreaker)
    results.sort(key=lambda x: (x['valid_count'], x['score'], x['distinct_attrs']), reverse=True)

    print(f"{'Location':<25} | {'Selection':<55} | {'Count':<5} | {'Score':<5} | {'Attr':<4} | Outcomes")
    print("-" * 140)
    # Only show top 100
    seen_selections = set()
    output_count = 0
    for r in results:
        sel_key = (r['location'], r['selection'])
        if sel_key in seen_selections:
            continue
        seen_selections.add(sel_key)
        
        outcomes_str = ", ".join([f"{n}(P{p})" for n, p in zip(r['outcomes'], r['priorities'])])
        print(f"{r['location']:<25} | {r['selection']:<55} | {r['valid_count']:<5} | {r['score']:<5} | {r['distinct_attrs']:<4} | {outcomes_str}")
        output_count += 1
        if output_count >= 100:
            break

if __name__ == '__main__':
    solve()

File: solver.js
// Engraving solver for Severe Energy Alluvium runs.
//
// An engraving picks 3 of the 5 attribute stats plus one fixed stat from the
// location's pool (a secondary stat or a skill stat). Each Flawless Essence
// then rolls one of the 3 picked attributes, carries the fixed stat, and rolls
// the remaining stat from the location's pool. A weapon can drop from an
// engraving when its attribute is picked, its fixed-side stat equals the fixed
// stat, and its other stat is in the location's pool. Weapons without a
// secondary stat (3-star) accept any secondary.

export const ATTRIBUTES = ['Strength', 'Agility', 'Will', 'Intellect', 'Main Attribute'];
export const TYPE_ORDER = ['Greatsword', 'Handcannon', 'Arts Unit', 'Sword', 'Polearm'];
export const PRIORITY = { SKIP: 0, TRACK: 1, STAR: 2 };
const ENGRAVED_ATTRIBUTES = 3;

export const hasSecondary = (w) => w.sec_stat !== 'None';

export function defaultPriority(w) {
  return w.rarity === 3 ? PRIORITY.SKIP : PRIORITY.TRACK;
}

// Returns the stat that must roll for weapon `w` to drop, '' when no roll is
// needed, or null when `w` cannot drop from this location and fixed stat.
export function requiredRoll(w, pool, fixed) {
  if (fixed.kind === 'secondary') {
    if (hasSecondary(w) && w.sec_stat !== fixed.stat) return null;
    return pool.skill.includes(w.skill_stat) ? w.skill_stat : null;
  }
  if (w.skill_stat !== fixed.stat) return null;
  if (!hasSecondary(w)) return '';
  return pool.secondary.includes(w.sec_stat) ? w.sec_stat : null;
}

function combinations(items, k) {
  if (k === 0) return [[]];
  if (items.length < k) return [];
  const [head, ...rest] = items;
  return [...combinations(rest, k - 1).map((c) => [head, ...c]), ...combinations(rest, k)];
}

const isStrictSubset = (a, b) => a.size < b.size && [...a].every((x) => b.has(x));

// Builds one row per distinct set of tracked weapons an engraving can yield.
// Engravings that differ only in untracked outcomes or in which fixed stat is
// chosen collapse into one row with several `options`. Within a location, a
// row whose tracked set is a strict subset of another row's is dropped. When
// any weapon is starred, only rows yielding a starred weapon are kept.
export function solve(data, priorities, { locations = null, limit = 100 } = {}) {
  const priorityOf = (w) => priorities[w.name] ?? defaultPriority(w);
  const anyStarred = data.weapons.some((w) => priorityOf(w) === PRIORITY.STAR);
  const locationNames = Object.keys(data.alluvium);
  const rows = [];

  for (const [location, pool] of Object.entries(data.alluvium)) {
    if (locations && !locations.has(location)) continue;
    const byHits = new Map();
    const fixedOptions = [
      ...pool.secondary.map((stat) => ({ kind: 'secondary', stat })),
      ...pool.skill.map((stat) => ({ kind: 'skill', stat })),
    ];

    for (const fixed of fixedOptions) {
      const eligible = [];
      for (const w of data.weapons) {
        const roll = requiredRoll(w, pool, fixed);
        if (roll !== null) eligible.push({ weapon: w, roll, priority: priorityOf(w) });
      }
      const tracked = eligible.filter((e) => e.priority > PRIORITY.SKIP);
      if (tracked.length === 0) continue;

      const trackedAttrs = ATTRIBUTES.filter((a) => tracked.some((e) => e.weapon.attr_stat === a));
      const attrSets = trackedAttrs.length <= ENGRAVED_ATTRIBUTES
        ? [trackedAttrs]
        : combinations(trackedAttrs, ENGRAVED_ATTRIBUTES);

      for (const attrs of attrSets) {
        const hits = tracked
          .filter((e) => attrs.includes(e.weapon.attr_stat))
          .sort((a, b) => b.priority - a.priority || b.weapon.rarity - a.weapon.rarity || a.weapon.name.localeCompare(b.weapon.name));
        const untracked = eligible
          .filter((e) => e.priority === PRIORITY.SKIP && attrs.includes(e.weapon.attr_stat))
          .map((e) => ({ weapon: e.weapon, roll: e.roll }));
        const key = hits.map((e) => e.weapon.name).sort().join('|');
        const option = { ...fixed, untracked };
        const row = byHits.get(key);
        if (row) {
          row.options.push(option);
        } else {
          byHits.set(key, {
            location,
            region: pool.region,
            attrs,
            freeAttrs: ENGRAVED_ATTRIBUTES - attrs.length,
            options: [option],
            hits: hits.map(({ weapon, priority }) => ({ weapon, priority, rollFor: {} })),
            starred: hits.filter((e) => e.priority === PRIORITY.STAR).length,
            tracked: hits.filter((e) => e.priority === PRIORITY.TRACK).length,
          });
        }
        // The stat a weapon still needs to roll depends on which option is engraved.
        const entry = byHits.get(key);
        for (const h of hits) entry.hits.find((x) => x.weapon === h.weapon).rollFor[fixed.stat] = h.roll;
      }
    }

    const candidates = [...byHits.values()];
    // Options yielding the same weapons can still differ in rolls: fixing a
    // 3-star weapon's skill leaves nothing else to roll. Keep only the options
    // that leave the most weapons needing no second roll.
    for (const row of candidates) {
      const noRoll = (o) => row.hits.filter((h) => h.rollFor[o.stat] === '').length;
      const best = Math.max(...row.options.map(noRoll));
      row.options = row.options.filter((o) => noRoll(o) === best);
    }
    const hitSets = candidates.map((r) => new Set(r.hits.map((h) => h.weapon.name)));
    candidates.forEach((row, i) => {
      if (!hitSets.some((other, j) => j !== i && isStrictSubset(hitSets[i], other))) rows.push(row);
    });
  }

  return rows
    .filter((r) => !anyStarred || r.starred > 0)
    .sort((a, b) =>
      b.starred - a.starred ||
      b.tracked - a.tracked ||
      b.freeAttrs - a.freeAttrs ||
      locationNames.indexOf(a.location) - locationNames.indexOf(b.location))
    .slice(0, limit);
}
